contracts/AllowedStrategy.sol

```diff
@@ -0,0 +1,270 @@
+
+        account = IAccount(_account);
+        groupHealthContract = IGroupHealth(_groupHealth);
+    }
+
+    /**
+     * @notice Marks a group as allowed strategy for voting.
+     * @param group The address of the group to add to the set of allowed
+     * strategies.
+     */
+    function allowStrategy(address group) external onlyManager {
+        if (!groupHealthContract.isValidGroup(group)) {
+            revert StrategyNotEligible(group);
+        }
+
+        if (allowedStrategies.contains(group)) {
+            revert StrategyAlreadyAdded(group);
+        }
+
+        if (!allowedStrategies.add(group)) {
+            revert FailedToAddAllowedStrategy(group);
+        }
+
+        emit StrategyAllowed(group);
+    }
+
+    /**
+     * @notice Marks a group as not allowed strategy for voting
+     * and redistributes votes to default strategy.
+     * @param group The address of the group to remove from the set of allowed
+     * strategies.
+     * @return total stCelo in blocked strategy
+     */
+    function blockStrategy(address group) external onlyManager returns (uint256) {
+        if (IManager(manager).getGroupsLength() == 0) {
+            revert NoActiveGroups();
+        }
+
+        if (!allowedStrategies.contains(group)) {
+            revert StrategyAlreadyBlocked(group);
+        }
+
+        if (!allowedStrategies.remove(group)) {
+            revert FailedToBlockStrategy(group);
+        }
+
+        emit StrategyBlocked(group);
+
+        return getTotalStCeloVotesForStrategy(group);
+    }
+
+    /**
+     * @notice Used to withdraw CELO from the system from allowed strategy
+     * that account voted for previously. It is expected that strategy will be balanced.
+     * For balancing use `rebalance` function
+     * @param strategy The validator group that we want to withdraw from.
+     * @param withdrawal The amount of stCELO to withdraw.
+     * @return groups The groups to withdraw from.
+     * @return votes The amount to withdraw from each group.
+     */
+    function withdrawFromAllowedStrategy(
+        address strategy,
+        uint256 withdrawal,
+        uint256 stCeloWithdrawalAmount
+    ) external onlyManager returns (address[] memory groups, uint256[] memory votes) {
+        uint256 votesRemaining = account.getCeloForGroup(strategy);
+        if (votesRemaining < withdrawal) {
+            revert GroupNotBalancedOrNotEnoughStCelo(strategy);
+        }
+
+        groups = new address[](1);
+        votes = new uint256[](1);
+        groups[0] = strategy;
+        votes[0] = withdrawal;
+
+        if (stCeloWithdrawalAmount > allowedStrategyTotalStCeloVotes[strategy]) {
+            revert CantWithdrawAccordingToStrategy(strategy);
+        }
+
+        allowedStrategyTotalStCeloVotes[strategy] -= stCeloWithdrawalAmount;
+        totalStCeloInAllowedStrategies -= stCeloWithdrawalAmount;
+    }
+
+    function addToTotalStCeloInAllowedStrategies(uint256 value) external onlyManager {
```

Comment:
Is there ever a case of this needing to be called without also calling `addToAllowedStrategyTotalStCeloVotes` for a specific strategy? If not, should this update just happen in `addToAllowedStrategyTotalStCeloVotes` to reduce complexity, make it easier to keep invariants, and reduce the number of calls to this contract?

Reply:
There is a case when transferring from one specific group strategy to other when we can just update totalStCelo in those specific strategies and not totalStCelo locked in all specific strategies, but it is similar to "toVote" "toRevoke" in Account - updated to one method only